src/c-unchop.c

```diff
@@ -9,48 +10,107 @@ enum fallback_homogeneous {
 #include "decl/c-unchop-decl.h"
 
 
-static
 r_obj* list_unchop(r_obj* xs,
                    r_obj* indices,
+                   r_obj* default_,
                    r_obj* ptype,
+                   struct optional_r_ssize size,
+                   enum list_unchop_unmatched unmatched,
                    r_obj* name_spec,
                    const struct name_repair_opts* name_repair,
                    struct vctrs_arg* p_error_arg,
+                   struct vctrs_arg* p_default_arg,
                    struct r_lazy error_call) {
   obj_check_list(xs, p_error_arg, error_call);
 
   if (indices == r_null) {
+    // Check if `default` and `size` were specified before reverting to `vec_c()`
+    if (default_ != r_null) {
+      r_abort("`indices` must be specified when `default` is specified.");
+    }
+    if (optional_r_ssize_is_some(size)) {
+      r_abort("`indices` must be specified when `size` is specified.");
+    }
+
     return vec_c(xs, ptype, name_spec, name_repair, p_error_arg, error_call);
```

Comment:
I have a feeling there is going to come a time where we give some effort to combine `vec_c()` and `list_unchop()` into a single code path. They share a lot of utilities right now but the differences are _just_ annoying enough that I think it would be better if we just shoved everything into one big function to think about.